### PR TITLE
catch `StateError` thrown from `Chromium.close`

### PR DESCRIPTION
**Pull request by andrewkolos**
Fixes crasher https://github.com/flutter/flutter/issues/154203.

Also does a little refactoring to move exception handling into the `getChromeTabGuarded` utility function.

I plan on cherry-picking a subset of this change. This would be included in the custom CP patch for https://github.com/flutter/flutter/issues/153064.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter_tools/lib/src/web/chrome.dart [hunk 1]
```diff
@@ -563,12 +564,23 @@ class Chromium {
...
   ChromeConnection chromeConnection,
   bool Function(ChromeTab tab) accept, {
   Duration? retryFor,
-}) {
-  return asyncGuard(() => chromeConnection.getTab(accept, retryFor: retryFor));
+  void Function(Object error, StackTrace stackTrace)? onError,
+}) async {
+  try {
+    return asyncGuard(() => chromeConnection.getTab(accept, retryFor: retryFor));
+  } on IOException catch (error, stackTrace) {
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/web/chrome.dart:575]: Maybe this should just be a blanket `on (e)` catch, but I'm being conservative.

File: packages/flutter_tools/lib/src/isolated/resident_web_runner.dart [hunk 2]
```diff
@@ -608,20 +608,17 @@ Please provide a valid TCP port (an integer between 0 and 65535, inclusive).
...
-        _logger.printError(error.toString(), stackTrace: stackTrace);
-        chromeTab = null;
-      }
+      final ChromeTab? chromeTab = await getChromeTabGuarded(
+        chrome.chromeConnection,
+        (ChromeTab chromeTab) {
+          return !chromeTab.url.startsWith('chrome-extension');
+        },
+        retryFor: const Duration(seconds: 5),
+        onError: (Object error, StackTrace stackTrace) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/isolated/resident_web_runner.dart:617]: does this now mean we're swallowing all errors?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/isolated/resident_web_runner.dart:617]: oh, it looks like this only gets delegated IOException and StateError?

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart [hunk 3]
```diff
@@ -454,9 +454,6 @@ class FlutterWebPlatform extends PlatformPlugin {
         } on FormatException catch (ex) {
           _logger.printError('Caught FormatException: $ex');
           return shelf.Response.ok('Caught exception: $ex');
-        } on IOException catch (ex) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/test/flutter_web_platform.dart:457]: Why are we not catching this here?

File: packages/flutter_tools/lib/src/isolated/resident_web_runner.dart (hunk 2 above)
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/isolated/resident_web_runner.dart:617]: oh yeah. Anything that's not `IOException` or `StateError` will fall through. Maybe it should just catch everything.

File: packages/flutter_tools/lib/src/test/flutter_web_platform.dart (hunk 3 above)
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/test/flutter_web_platform.dart:457]: This catch was introduced recently, but before I introduced `getChromeTabGuarded`. `getChromeTabGuarded` should already catch this.

File: packages/flutter_tools/lib/src/web/chrome.dart [hunk 4]
```diff
@@ -563,12 +564,23 @@ class Chromium {
...
+    if (onError != null) {
+      onError(error, stackTrace);
+    }
+    return null;
+  } on StateError catch (error, stackTrace) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/web/chrome.dart:580]: can you add a comment why we're catching StateError?

File: packages/flutter_tools/lib/src/web/chrome.dart [hunk 5]
```diff
@@ -563,12 +564,23 @@ class Chromium {
   }
 }
 
-// TODO(andrewkolos): Remove when https://github.com/dart-lang/sdk/issues/56566
-//  is fixed.
 Future<ChromeTab?> getChromeTabGuarded(
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/web/chrome.dart:567]: Can you add a dartdoc for how to use this function?

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/web/chrome.dart:574]: Addressed in [refactor: add clarifying docs](https://github.com/flutter/flutter/pull/154366/commits/9fd3e846b7d2cb8761637839cbca769d38e3f0f3)

File: packages/flutter_tools/lib/src/web/chrome.dart (hunk 4 above)
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/web/chrome.dart:594]: Addressed in [refactor: add clarifying docs](https://github.com/flutter/flutter/pull/154366/commits/9fd3e846b7d2cb8761637839cbca769d38e3f0f3)

File: packages/flutter_tools/lib/src/isolated/resident_web_runner.dart (hunk 2 above)
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/isolated/resident_web_runner.dart]: Addressed in [refactor: add clarifying docs](https://github.com/flutter/flutter/pull/154366/commits/9fd3e846b7d2cb8761637839cbca769d38e3f0f3). I decided to just rename the parameter to not imply that any error would be caught.

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**